drivers/cleanup.py

```diff
@@ -1579,6 +1582,48 @@ def __init__(self, uuid, xapi, createLock, force):
         elif not self.xapi.isMaster():
             raise util.SMException("This host is NOT master, will not run")
 
+        self.no_space_candidates = {}
+
+    def msg_cleared(self, xapi_session, msg_ref):
+        msg = None
+        try:
+            msg = xapi_session.xenapi.message.get_record(msg_ref)
+        except XenAPI.Failure:
+            return True
+
+        return msg is None
+
+    def check_no_space_candidates(self):
+        xapi_session = self.xapi.getSession()
+        for candidate in self.no_space_candidates.values():
+            msg_id = candidate.getConfig(VDI.DB_GC_NO_SPACE)
+            if msg_id is None or self.msg_cleared(xapi_session, msg_id):
+                util.SMlog(f"Could not coalesce {candidate} "
+                           "due to a lack of space in the SR")
+                msg_body = ("Unable to perform data coalesce on vdi "
+                            f"{candidate.uuid} due to a lack of space in SR "
+                            f"{self.uuid}")
+                msg_id = xapi_session.xenapi.message.create(
+                    'SM_GC_NO_SPACE',
+                    3,
+                    "VDI",
+                    candidate.uuid,
+                    msg_body)
+                candidate.setConfig(VDI.DB_GC_NO_SPACE, msg_id)
+
+    def clear_no_space_msg(self, vdi):
+        xapi_session = self.xapi.getSession()
+        msg_id = None
+        try:
+            msg_id = vdi.getConfig(VDI.DB_GC_NO_SPACE)
```

Comment:
Are you sure you want to store this per VDI? If you have 1000 VDIs, then as soon as the SR runs out of space you'd get flooded with a 1000 messages (even if the GC stops at the first one, is it guaranteed to always iterate on VDIs in the same order?)
Could this be stored per PBD or SR instead?

Reply:
Yes, it need to be stored per VDI. As you say some may coalesce, others may not.

Reply:
Maybe what it needs to do is raise one message to Xapi and then tag that ID to all the impacted VDIs, then remove the tag if that VDI become coalescable in the future. Finally removing the message from the SR if we complete a GC run with no VDIs unable to coalesce.